packages/Search/src/DTK_FineSearchFunctor.hpp

```diff
@@ -0,0 +1,361 @@
+                         double threshold ) const
+    {
+        return Intrepid2::Impl::Wedge<18>::checkPointInclusion( reference_point,
+                                                                threshold );
+    }
+};
+}
+
+namespace Functor
+{
+template <typename CellType, typename DeviceType>
+class FineSearch
+{
+  public:
+    FineSearch(
+        Kokkos::View<Coordinate **, DeviceType> reference_points,
+        Kokkos::View<bool *, DeviceType> point_in_cell,
+        Kokkos::View<Coordinate **, DeviceType> physical_points,
+        Kokkos::View<Coordinate ***, DeviceType> cells,
+        Kokkos::View<unsigned int *, DeviceType> coarse_search_output_points,
+        Kokkos::View<unsigned int *, DeviceType> coarse_search_output_cells );
+
+    KOKKOS_INLINE_FUNCTION
+    void operator()( unsigned int const i ) const
+    {
+        unsigned int const point_index = _coarse_search_output_points( i );
+        unsigned int const cell_index = _coarse_search_output_cells( i );
+        // Get subviews
+        Kokkos::View<Coordinate *, DeviceType> ref_point( _reference_points, i,
+                                                          Kokkos::ALL() );
+        Kokkos::View<Coordinate *, DeviceType> phys_point(
+            _physical_points, point_index, Kokkos::ALL() );
+        Kokkos::View<Coordinate **, DeviceType> nodes(
+            _cells, cell_index, Kokkos::ALL(), Kokkos::ALL() );
+
+        // Compute the reference point and return flag for the presence of the
+        // point inside the cell
+        _cell_type.mapToReferenceFrame( ref_point, phys_point, nodes );
+        _point_in_cell[i] =
+            _cell_type.checkPointInclusion( ref_point, _threshold );
+    }
+
+  private:
+    CellType _cell_type;
+    double _threshold;
+    Kokkos::View<Coordinate **, DeviceType> _reference_points;
+    Kokkos::View<bool *, DeviceType> _point_in_cell;
+    Kokkos::View<Coordinate **, DeviceType> _physical_points;
+    Kokkos::View<Coordinate ***, DeviceType> _cells;
+    Kokkos::View<unsigned int *, DeviceType> _coarse_search_output_points;
+    Kokkos::View<unsigned int *, DeviceType> _coarse_search_output_cells;
+};
+
+template <typename CellType, typename DeviceType>
+FineSearch<CellType, DeviceType>::FineSearch(
+    Kokkos::View<Coordinate **, DeviceType> reference_points,
+    Kokkos::View<bool *, DeviceType> point_in_cell,
+    Kokkos::View<Coordinate **, DeviceType> physical_points,
+    Kokkos::View<Coordinate ***, DeviceType> cells,
+    Kokkos::View<unsigned int *, DeviceType> coarse_search_output_points,
+    Kokkos::View<unsigned int *, DeviceType> coarse_search_output_cells )
+    : _threshold( 0. )
+    , _reference_points( reference_points )
+    , _point_in_cell( point_in_cell )
+    , _physical_points( physical_points )
+    , _cells( cells )
+    , _coarse_search_output_points( coarse_search_output_points )
+    , _coarse_search_output_cells( coarse_search_output_cells )
+{
+}
```

Comment:
Add a check that the sizes of `_coarse_search_output_points` and `_coarse_search_output_cells` are the same.